File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/H-Наибольшая_возрастающая_подпоследовательность/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(N^2)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<int> nums (n, 0);
    std::vector<int> prev (n, -1);
    for (int i = 0; i < n; ++i) std::scanf("%d", &nums[i]);
    std::vector<int> dp(n, 1);
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < i; ++j) {
            if (nums[i] > nums[j] && dp[i] < dp[j] + 1) {
                dp[i] = dp[j] + 1;
                prev[i] = j;
            }
        }
    }
    int i = 0;
    for (int j = 1; j < n; ++j) {
        if (dp[i] < dp[j]) i = j;
    }
    std::vector<int> path;
    path.push_back(nums[i]);
    while (prev[i] != -1) {
        path.push_back(nums[prev[i]]);
        i = prev[i];
    }
    std::printf("%d\n", path.size());
    for (int j = path.size() - 1; j >= 0; --j) {
        std::printf("%d ", path[j]);
    }
    std::printf("\n");
    
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/H-Наибольшая_возрастающая_подпоследовательность/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N^2)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    prev: list[int] = [-1] * n
    dp: list[int] = [1] * n
    for i in range(n):
        for j in range(i):
            if nums[i] > nums[j] and dp[i] < dp[j] + 1:
                dp[i] = dp[j] + 1
                prev[i] = j
    i: int = 0
    for j in range(1, n):
        if dp[i] < dp[j]: i = j
    path: list[int] = [nums[i]]
    while prev[i] != -1:
        path.append(nums[prev[i]])
        i = prev[i]
    sys.stdout.write('{}\n'.format(len(path)))
    sys.stdout.write('{}\n'.format(' '.join(str(move) for move in reversed(path))))


if __name__ == '__main__':
    solution()